### PR TITLE
IMDB Reviews dataset

### PR DESCRIPTION
**Pull request by t-rutten**
These changes add training and test sets for the [IMDB reviews dataset](http://ai.stanford.edu/~amaas/data/sentiment/) related to #11. Train and test sets each contain 25000 labeled examples; each example includes a textual review of a movie and a binary label categorizing the example as having positive sentiment (1) or negative sentiment (0).

The dataset can be queried like:

```elixir
{inputs, targets} = Scidata.IMDBReviews.download
{test_inputs, test_targets} = Scidata.IMDBReviews.download_test
```

### Questions and Discussion
-  The source dataset also includes 50000 unlabeled examples useful for unsupervised learning. Do we want to add that set to the API as well?
- The return tuple for `download/1` and `download_test/1` is two lists, but we could return raw binaries instead (where inputs might be separated by newline character). 
- We don't provide a type or a shape for the return values--there's no tensor type associated with the data, and the "shape" of each example input is variable because reviews have variable length. Is there type that seems sensible to provide in the return tuple? The shape we might specify as `{25000, nil}` for inputs and `{25000}` for labels.
- We could make return labels and datasets as streams for faster initial load and to avoid keeping it in memory, but the dataset is not massive so I'm unsure if it's worth it.
- It seems some codepoints aren't UTF-8 encoded and thus not "printable". This leads to some examples like
```elixir
[..., 
   "...The BBC version is so superior it's not even funny and everything about this version is an insult to its memory. In short if you must see it be sure you have read the book first or seen the BBC version other wise you will be lead done the deluded road that this is what it's like, which its not!",
   <<75, 97, 122, 117, 111, 32, 75, 111, 109, 105, 122, 117, 32, 115, 116, 114,
     105, 107, 101, 115, 32, 97, 103, 97, 105, 110, 32, 119, 105, 116, 104, 32,
     34, 69, 110, 116, 114, 97, 105, 108, 115, ...>>,
   "Daniel Percival's \"Dirty War\", a BBC production made for television was shown recently on cable. The film has a documentary style in the way it goes after the people that caused the near holocaust in one of the big metropolis of the world, London...",
...]
```
We probably don't need to address this—users can clean and tokenize bitstrings of each entry as desired—but I wanted to flag. The experience on [TFDS](https://www.tensorflow.org/datasets/catalog/imdb_reviews) is similar.

### CODE REVIEW

**Comment by seanmor5** [comment]
@t-rutten I think I addressed most of the discussion points in the issue I just opened. As far as different sequence lengths, we'll have to pad to max length or truncate to a fixed length to batch as we don't support ragged tensors. I think leaving everything as a list is okay for now.

**Comment by t-rutten** [comment]
Great @seanmor5, I'll have a look.

File: lib/scidata/imdb_reviews.ex [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      Enum.filter(files, fn {fname, _} ->
+        file_match?(fname, dataset_type, :pos)
+      end)
+
+    neg_files =
+      Enum.filter(files, fn {fname, _} ->
+        file_match?(fname, dataset_type, :neg)
+      end)
+
+    :rand.seed(:exsss, {101, 102, 103})
```
Comment (josevalim) [inline lib/scidata/imdb_reviews.ex]: If we are hardcoding a seed, then shuffle is deterministic. Is that what we really want? Also, maybe we should let the users shuffle?

Also, should we let positive and negative be an argument... or not prefilter by postive and negative? I think we should try to avoid doing passes over the data, as that can be expensive?

Comment (t-rutten) [reply lib/scidata/imdb_reviews.ex]: Deterministic shuffling is probably not what we want, and if we shuffle the dataset then ideally the user should pass a seed (seed can be an opt in a future dataset pipeline #13). I like your suggestion to let the user shuffle for now, let's go with that.

If a user wants only labeled examples for supervised learning and we don't filter the files in the positive and negative directories, then they'll load 50000 examples that'll be unused. I think it would be reasonable to let users specify which kinds of examples they want (positive, negative, unlabeled) in an argument--I'll add that.

File: lib/scidata/imdb_reviews.ex [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      files
+      |> Enum.filter(fn {fname, _} ->
+        file_match?(fname, dataset_type, example_types)
+      end)
+      |> Enum.reduce(
+        {[], []},
+        fn {fname, contents}, {inputs, labels} ->
+          {[contents | inputs], [get_label(fname) | labels]}
+        end
+      )
```
Comment (josevalim) [inline lib/scidata/imdb_reviews.ex]: You can also try this for a single pass on the data :)

```suggestion
      for {fname, contents} <- files,
          file_match?(fname, dataset_type, example_types),
          reduce: {[], []} do
        {inputs, labels} ->
          {[contents | inputs], [get_label(fname) | labels]}
      end)
```

**Review by josevalim — COMMENTED** [review]
I dropped one comment on the Elixir code :) Looking great!

**Comment by t-rutten** [comment]
@seanmor5 do you still think the list version of the data set here is sufficient, or should we change to a truncated or padded tensor instead?

**Comment by josevalim** [comment]
According to our discussions on Slack, some datasets may not be suitable for tensors, therefore we should have a different return type. I think this one could be suitable for dataframes, so we could return the data in this format:

```elixir
%{
  review: [review1, review2, review3, ...]
  sentiment: [1, -1, 0, ...]
}
```

All datasets should likely live in a flat namespace and we will use ExDoc grouping to categorize them and also use specs to clarify the return type (series, tensors, etc).

**Comment by t-rutten** [comment]
Thanks @josevalim! I will add changes to the returned data.

**Comment by t-rutten** [comment]
@josevalim I've changed the return types and added specs for `download/2` and `download_test/2`. Thanks for reviewing!

File: lib/scidata/imdb_reviews.ex [hunk 3]
```diff
@@ -28,6 +36,8 @@ defmodule Scidata.IMDBReviews do
   `example_types` is the same argument in `download/2` but excludes `:unsup`
   because all unlabeled examples are in the training set.
   """
+  @spec download_test(sentiments, opts()) ::
+          %{review: any(), sentiment: any()}
```
Comment (josevalim) [inline lib/scidata/imdb_reviews.ex]: What if we document the default return types by documenting the behaviour with no opts? Something like:


```suggestion
  @spec download_test(sentiments) :: %{review: binary(), sentiment: -1 | 1}
  @spec download_test(sentiments, opts()) :: %{review: any(), sentiment: any()}
```

File: lib/scidata/imdb_reviews.ex [hunk 4]
```diff
@@ -8,13 +8,21 @@ defmodule Scidata.IMDBReviews do
 
   alias Scidata.Utils
 
+  @type sentiments :: [atom]
+  @type transform_type :: atom
+  @type transform_fn :: ([binary, ...] -> any)
+  @type transform_opt :: {transform_type, transform_fn}
+  @type opts :: [transform_opt]
```
Comment (josevalim) [inline lib/scidata/imdb_reviews.ex]: I think we can explicitly document the options. Something like this:

```suggestion
  @type opts :: [
    transform_inputs: (...precisetype...),
    transform_labels: (...precisetype...)
  ]
```

File: lib/scidata/imdb_reviews.ex (hunk 3 above)
Comment (josevalim) [reply lib/scidata/imdb_reviews.ex]: We need to apply a similar change above to `download`. :)

Also, can we have unsup on both `download` and `download_test`?

Comment (t-rutten) [reply lib/scidata/imdb_reviews.ex]: @josevalim good catch! `:unsup` only belongs with `download`.

**Comment by t-rutten** [comment]
@josevalim let me know if I missed anything!
- `download` and `download_test` now have specs which show argument and return types without `opts`
- `opts` spec is more explicit

File: lib/scidata/imdb_reviews.ex [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+  defp file_match?(fname, dataset_type, example_types) do
+    pattern = ~r/#{dataset_type}\/(#{Enum.join(example_types, "|")})\//
+    String.match?(List.to_string(fname), pattern)
+  end
+
+  defp get_label(fname) do
+    cond do
+      String.match?(List.to_string(fname), ~r/pos/) -> 1
+      String.match?(List.to_string(fname), ~r/neg/) -> 0
```
Comment (josevalim) [inline lib/scidata/imdb_reviews.ex]: Should this be -1?

File: lib/scidata/imdb_reviews.ex [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+          {[contents | inputs], [get_label(fname) | labels]}
+      end
+
+    %{review: transform_inputs.(inputs), sentiment: transform_labels.(labels)}
+  end
+
+  defp file_match?(fname, dataset_type, example_types) do
+    pattern = ~r/#{dataset_type}\/(#{Enum.join(example_types, "|")})\//
+    String.match?(List.to_string(fname), pattern)
+  end
```
Comment (josevalim) [inline lib/scidata/imdb_reviews.ex]: Compile regex only once:

```suggestion
    files = Utils.get!(@base_url <> @dataset_file).body
    regex = ~r"#{dataset_type}/(#{Enum.join(example_types, "|")})/"

    {inputs, labels} =
      for {fname, contents} <- files,
          List.to_string(fname) =~ regex,
          reduce: {[], []} do
        {inputs, labels} ->
          {[contents | inputs], [get_label(fname) | labels]}
      end

    %{review: transform_inputs.(inputs), sentiment: transform_labels.(labels)}
  end
```

:)

File: lib/scidata/imdb_reviews.ex [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+  defp get_label(fname) do
+    cond do
+      String.match?(List.to_string(fname), ~r/pos/) -> 1
+      String.match?(List.to_string(fname), ~r/neg/) -> 0
+      String.match?(List.to_string(fname), ~r/unsup/) -> nil
+    end
```
Comment (josevalim) [inline lib/scidata/imdb_reviews.ex]: Convert to string once and use binary matching instead of regex:

```suggestion
    fname = List.to_string(fname)

    cond do
      fname =~ "pos" -> 1
      fname =~ "neg" -> -1
      fname =~ "unsup" -> nil
    end
```

File: lib/scidata/imdb_reviews.ex [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  @doc """
+  Downloads the IMDB reviews training dataset or fetches it locally.
+
+  `example_types` specifies which examples in the dataset should be returned
+  according to each example's label: `:pos` for positive examples, `:neg` for
+  negative examples, and `:unsup` for unlabeled examples.
+  """
+  @spec download([train_sentiment]) :: %{review: [binary(), ...], sentiment: 1 | -1}
+  def download(
+        example_types \\ [:pos, :neg],
```
Comment (josevalim) [inline lib/scidata/imdb_reviews.ex]: Maybe we should make the example types an option to simplify the API? 🤔 For example:

    download(example_types: [:pos])

**Review by josevalim — COMMENTED** [review]
I drop another round of reviews, sorry for the back and forth, but I should have probably dropped this one the first time. :)

File: lib/scidata/imdb_reviews.ex (hunk 5 above)
Comment (t-rutten) [reply lib/scidata/imdb_reviews.ex]: It should be -1 according to spec now. TFDS labels negative sentiment examples as 0 (see "Display examples..." [here](https://www.tensorflow.org/datasets/catalog/imdb_reviews#imdb_reviewsplain_text_default_config)).

Would you prefer -1 or 0? I'll make sure the label and spec correspond.

Comment (josevalim) [reply lib/scidata/imdb_reviews.ex]: Good point. Actually, let’s go with 0, because then we can represent it with u8 or predicate types.

File: lib/scidata/imdb_reviews.ex (hunk 8 above)
Comment (t-rutten) [reply lib/scidata/imdb_reviews.ex]: Simpler API is good! Something like this?

```elixir
  @spec download(example_types: [test_sentiment]) :: %{review: [binary(), ...], sentiment: 1 | 0}
  def download(
        opts \\ [example_types: [:pos, :neg]]
      ) do
    {example_types, opts} = Keyword.pop(opts, :example_types)
    download_dataset(example_types, :train, opts)
  end
```

Comment (josevalim) [reply lib/scidata/imdb_reviews.ex]: Yes! Although I think you can handle the options inside download_dataset:

    download_dataset(:train, opts)

And in there:

    example_types = opts[:example_types] || [:pos, :neg]

Comment (t-rutten) [reply lib/scidata/imdb_reviews.ex]: @josevalim good call, I've added those changes.

**Review by josevalim — APPROVED** [review]
